rfcs/0039-minaform-tool-design.md

```diff
@@ -0,0 +1,143 @@
+
+This command can be outsourced to google cloud for expediency, and because these files including the image itself never really need to be on the same machine that operates the deployment. To this end, google cloud build (https://cloud.google.com/cloud-build/docs/quickstart-build#build_using_dockerfile) a serverless product that can be sent a dockerfile at any time to have google build it. There is some value in trying to handle other CI builds with google cloud build but its especially perfect for this scenario when the majority of the build time is spent on network transfers to/from google cloud, so that GCB can build much quicker than any other environment (and we pay per build minute over 120/day). For nearly 0 additional development effort we get to bake docker images in google cloud.
+
+### Minaform Start
+
+The start command handles deploying the network and starting all containers / gcloud resources. `minaform start <testnet_name>` will start the given testnet on the default cluster, and provide a output to display the deployment progress. Keys will be uploaded to the cluster when the namespace is created, and then terraform resources will be deployed. If `--wait` is specified, the script will run `minaform watch <testnet_name>` once terraform is complete to watch for all of the nodes to become ready.
+
+### Minaform Stop
+
+The stop command handles destroying all resources for the given testnet, ideally using the existing terraform state and all of the same pararemeters so destroys can always happen smoothly. Will provide a `--force` flag to force terraform and other tooling to delete everything and clean up. This command closely matching the process for start/restart/etc. addresses pain point #3 and should make it easier to keep the cluster clean and easy.
+
+### Minaform Restart
+
+The restart command runs a `minaform stop` followed by `minaform start` and accepts the same flags as both start + stop.
+
+### Minaform Upgrade
+
+The upgrade command runs a terraform apply, to update the existing deployment with new changes. Ideally this should also provide more debugging flags to terraform and other tooling, and require user input to proceed so that the engineer deploying can understand what state will be lost (if any) in the process. Accepts the same flags as start/stop/restart.
+
+### Minaform connect
+
+To facilitate easy local testing, `minaform connect <testnet_name>` will use the terraform.env to locally start a docker container node that connects to the deployed testnet. With some additional flags to specify keypairs to mount `--key <key folder path>` and where to persist the .coda-config dir (default nowhere) `--local-config-dir <path>`. This is another area where many developers know some process for assembling the correct docker image, daemon.json, and peers for a given network but with the terraform.env and kubectl its very possible to script and lower the barrier to entry even more.
```

Comment:
A dedicated docker image containing tooling for inspecting/interacting with testnets sounds good though feels like we could just launch an actual pod or job pod within a testnet namespace, using this image, and `(terraform) output` the command to `exec` in whenever triggered. This way not only is a developer's  debug/inspection state maintained regardless of what he/she chooses to do with their laptop while enabling simpler to reason about pod/cluster network interactions, but also allows for greater range in sizing for more resource intense tasks.